packages/framework-core/test/booster-auth.test.ts

```diff
@@ -250,3 +253,63 @@ describe('the "isUserAuthorized" method', () => {
     expect(BoosterAuth.isUserAuthorized(authorizedRoles, userEnvelope)).to.eq(true)
   })
 })
+
+describe('the "verifyToken" method', () => {
+  const auth0VerifierUri = 'https://myauth0app.auth0.com/'
+  const issuer = 'auth0'
+  const jwks = createJWKSMock(auth0VerifierUri)
+  const email = internet.email()
+  const phoneNumber = phone.phoneNumber()
+  const userId = random.uuid()
+  const config = new BoosterConfig('test')
+  config.tokenVerifier = {
+    issuer,
+    jwksUri: auth0VerifierUri + '.well-known/jwks.json',
+  }
+
+  beforeEach(() => {
+    jwks.start()
+  })
+
+  afterEach(async () => {
+    await jwks.stop()
+  })
+
+  it('decode and verify an auth token with the custom roles', async () => {
+    const token = jwks.token({
+      sub: userId,
+      iss: issuer,
+      'custom:role': 'User',
```

Comment:
Exactly. @charlietfe and I give this exact same concept a twist and thought about using a function in the `auth` field, so that the user can do whatever they want, not just comparing claims as strings.
Among other reasons, it is perfectly possible (Hasura does this) that one of the claims is an array of possible roles, so to check the authorization we should see if the user role is inside that array of roles.

Because of these corner cases, using a function is much more flexible. For example:
```typescript
@Command({
  auth: (userEnvelope: UserEnvelope): boolean => { // The UserEnvelop has the claims extracted from the token
    return ['writeAll', 'writePayments'].contains(userEnvelope.claims['permission']) // For example
  }
})
export class ConfirmPayment {
  ...
}
```

This way, Booster now forgets completely about the authorization part too (including roles!) making this ultra-decoupled.

However, this could be clunky for people (a lambda there and repeated multiple times). So the idea is to provide the same behavior we have now with Booster provided functions. This is an example of your example but with a function that just does a straight comparison:

```typescript
@Command({
  auth: HasClaims({
    role: 'User', // The token includes a 'custom:role': 'User' claim
    permission: 'FileAccess', // The token includes a 'custom:permission': 'FileAccess' claim
    origin: 'CartService', // The token includes a 'custom:origin': 'CartService' claim
  })
})
export class ConfirmPayment {
  ...
}
```
Finally, as we are going to move the current authentication feature to a rocket, _it is that rocket the one that could use a role-based authentication_. For example, imagine we are using the canonical `BoosterAuth` rocket, which contains exactly the same we have right now in the core. The example would then look like:
```typescript
@Command({
  auth: BoosterAuth.hasRole('Admin')
})
export class ConfirmPayment {
  ...
}
```

💥  So we get a lot of flexibility maintaining practically the same user experience!!

@javiertoledo @charlietfe Let me know what you think

Reply:
Flexibility is always great, but a flexible interface comes with extra decisions for the final user and less-readable DSLs: you have to understand that you need to put a function there, what it should do, and what's the catalog of available "shortcut" functions (Like `HasRoles`), while a plain type could be autocompleted by the IDE. While this looks like a minor detail, it can impact the way people learn to use the framework, so I'd give extra thought to this DSL. One of Booster's principles is to minimize the amount of "code" and config required for common tasks. I love the idea of having a "fallback function" that is run when no authorization parameters are passed, but putting that function in a decorator parameter sounds weird; it's not usually where people expect to put code... I'd propose supporting the following scenarios:

1. Lower-level claim-based authorization: Requires all the described claims to be present and match. In this case, claims semantics are defined in the Auth provider.
```typescript
@Command({
  requireClaims: {
    role: 'User',
    permission: 'FileAccess',
    origin: 'CartService',
  }
})
export class ConfirmPayment {
  ...
}
```

2. Implement an `authorize` method in the command class where the action can be canceled by throwing an exception, or the token can be extended before injecting it in the command:
```typescript
@Comand
export class ConfirmPayment {
  public static function authorize(jwtToken: JWTToken, command: ConfirmPayment): JWTToken {
    // Maybe we expect a special claim in this command that includes a message ciphered with a secret from the payment provider
    verifyAuthenticityOfPaymentProvider(jwtToken.claims['evil-corp:creepy-secret'])
    return jwtToken
  }
  
  public static function handle(...) {}
}
```

3. It's important to mention that a command where neither one of these DSLs or the `authorize` method is defined should be inaccessible. Booster should even avoid creating the entries in the GraphQL schema. So it'd be also convenient to provide a special shortcut to make them publicly accessible:
```typescript
@Command({
  public: true
})
export class CreateUser {
  public static function handle(...) {}
}
```

With these features we should cover most scenarios. In the future, we can provide extra shortcuts for common cases like authorization via roles or via permissions:

```typescript
@Command({
  requireRole: 'Admin',
})
export class CreateUser {
  public static function handle(...) {}
}
```

Reply:
@javiertoledo  @alvaroloes After giving a thought, I'm more convinced that we should avoid functions as a part of the DSL definition, why? Because it will let the developers to create crazy functions involving even async operations to check i.e a role with a 3rd party provider or something like that, and for me it doesn't feel natural. 

Is better to have a well defined DSL which internally will call the default `authorize` method checking the required roles or on the contrary overwrite `authorize` method and provide your own logic if you want.

We are still covering all the initial use cases we had in mind with a lot of flexibility, also if you want to still create helper functions inside the framework to call them inside `authorize` you can do it as well, but without poluting the DSL with code. Anyway the problem with that approach is how the user will find those helper functions, and even more we should add more documentation to specify how to use them.